format-common/src/main/java/io/cdap/plugin/format/charset/fixedlength/FixedLengthCharsetTransformingDecompressor.java

```diff
@@ -0,0 +1,266 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(FixedLengthCharsetTransformingDecompressor.class);
+
+  protected final FixedLengthCharset sourceEncoding;
+  protected final CharsetDecoder decoder;
+  protected final CharsetEncoder encoder;
+  protected final Charset targetCharset = StandardCharsets.UTF_8;
+  protected long numDecodedCharacters = 0;
+  protected long numEncodedCharacters = 0;
+
+  //Initializing all buffers.
+  protected ByteBuffer inputByteBuffer = ByteBuffer.allocate(0);
+  protected CharBuffer decodedCharBuffer = CharBuffer.allocate(0);
+  protected ByteBuffer partialOutputByteBuffer = ByteBuffer.allocate(0);
+
+  public FixedLengthCharsetTransformingDecompressor(FixedLengthCharset sourceEncoding) {
+    this.sourceEncoding = sourceEncoding;
+    this.decoder = sourceEncoding.getCharset().newDecoder();
+    this.encoder = targetCharset.newEncoder();
+  }
+
+  @Override
+  public void setInput(byte[] b, int off, int len) {
+    //Expand incoming buffer if needed.
+    if (inputByteBuffer.remaining() < len) {
+      //Allocate new buffer that can fill the existing input + newly received bytes
+      ByteBuffer newIncomingBuffer = ByteBuffer.allocate(len + inputByteBuffer.capacity());
+
+      //Set up incoming buffer for reads and copy contents into new buffer.
+      inputByteBuffer.flip();
+      newIncomingBuffer.put(inputByteBuffer);
+
+      inputByteBuffer = newIncomingBuffer;
+    }
+
+    //Copy incoming payload into Input Byte Buffer
+    inputByteBuffer.put(b, off, len);
+    inputByteBuffer.flip();
+
+    //Set up char buffer for writes
+    decodedCharBuffer.compact();
+
+    //Expand the char buffer if needed.
+    if (decodedCharBuffer.capacity() < inputByteBuffer.limit() / sourceEncoding.getNumBytesPerCharacter()) {
+      decodedCharBuffer = CharBuffer.allocate(inputByteBuffer.limit() / sourceEncoding.getNumBytesPerCharacter());
+    }
+
+    //Decode bytes from the input buffer into the Decoded Char Buffer
+    decodeByteBufferIntoCharBuffer(inputByteBuffer);
+
+    //Set up decoded char buffer for reads.
+    decodedCharBuffer.flip();
+
+    //Set up incoming buffer for writes.
+    inputByteBuffer.compact();
+
+  }
+
+  /**
+   * Note that we only ask for additional input once we have completely depleted out outgoing buffer.
+   */
+  @Override
+  public boolean needsInput() {
+    return decodedCharBuffer.remaining() == 0;
+  }
+
+  @Override
+  public void setDictionary(byte[] b, int off, int len) {
+    //no-op
+  }
+
+  @Override
+  public boolean needsDictionary() {
+    return false;
+  }
+
+  @Override
+  public boolean finished() {
+    return decodedCharBuffer.remaining() == 0 && partialOutputByteBuffer.remaining() == 0;
+  }
+
+  @Override
+  public int decompress(byte[] b, int off, int len) throws IOException {
+
+    //Allocate new outgoing buffer
+    ByteBuffer encodedBuffer = ByteBuffer.allocate(len - off);
```

Comment:
Can we just do `ByteBuffer.wrap(b, off, len)` and save one more copy?

Reply:
Good catch. I'll fix this.